diff --git a/src/game/ingame_menu.c b/src/game/ingame_menu.c
index b9a43df..4006a4b 100644
--- a/src/game/ingame_menu.c
+++ b/src/game/ingame_menu.c
@@ -2765,7 +2765,7 @@ void print_hud_course_complete_coins(s16 x, s16 y) {
             gCourseCompleteCoins++;
             play_sound(SOUND_MENU_YOSHI_GAIN_LIVES, gDefaultSoundArgs);
 
-            if (gCourseCompleteCoins == 50 || gCourseCompleteCoins == 100 || gCourseCompleteCoins == 150) {
+            if (gCourseCompleteCoins % 50 == 0) {
                 play_sound(SOUND_GENERAL_COLLECT_1UP, gDefaultSoundArgs);
                 gMarioState[0].numLives++;
             }
